### PR TITLE
Refactor TWPCA API (work in progress)

### PR DESCRIPTION
**Pull request by ahwillia**
The goal of this PR is to (a) simplify model initialization by pulling the helper functions in `warp.py` and `utils.py` into the main TWPCA class, and (b) give the user more flexibility in fitting and initializing the model. In particular, the functions `generate_warps` and `compute_low_rank_factors` were beastly and referenced a ton of options that were internal to the TWPCA class anyways. They were not really exposed to the user.

Also, the initialization for the factors and the warps are now decoupled from the `fit` function. Check out `TWPCA.assign_warps(...)` and `TWPCA.assign_factors(...)`. The user can now specify the warps directly - which is nice because e.g. you can specify warping functions that map from clock time to an animal's position on a 1D track.

I've also renamed variables to be more consistent. Pretty much everything with an underscore before it is a tensorflow variable, while the variable without the underscore is a numpy array.

I'd like to nail down a better option for tensorflow session management before merging. At the moment, this breaks the cross-validation code.

### CODE REVIEW

File: twpca/model.py [hunk 1]
```diff
@@ -55,26 +86,154 @@ def __init__(self, n_components, shared_length=None,
...
+        if warps is not None:
+            # warps proved by user
+            if normalize_warps:
+                warps *= self.n_timepoints / np.max(warps)
+
+            shift = warps[:, 0] - 1
+            scale = np.ones(self.n_trials)
+            dtau = np.hstack((np.ones((self.n_trials, 1)), np.diff(warps, axis=1)))
+            tau = utils.inverse_softplus(np.maximum(0, dtau) * np.log(2.0))
+
```
Comment (nirum) [inline twpca/model.py:266]: add back 'identity'

File: twpca/model.py [hunk 2]
```diff
@@ -38,13 +42,40 @@ def __init__(self, n_components, shared_length=None,
             origin_idx (optional): if not None, all warping functions are pinned (aligned) at this
                 index. (default: None)
         """
+
```
Comment (nirum) [inline twpca/model.py]: create a separate graph in init and store it on self

File: twpca/model.py [hunk 3]
```diff
@@ -55,26 +86,154 @@ def __init__(self, n_components, shared_length=None,
         }
```
Comment (nirum) [inline twpca/model.py:105]: put code that creates tensorflow variables in a `with` block using `self._graph` as default

File: twpca/model.py [hunk 4]
```diff
@@ -247,12 +340,67 @@ def predict(self, X=None):
...
+
+        # store the warping function
+        self._params['warp'] = _warp
+
+        # store the inverse warping function
+        _args = [_warp, self.n_timepoints, self.shared_length]
+        self._inv_warp = tf.py_func(warp._invert_warp_indices, _args, tf.float32)
+
+        # declare which parameters are trainable
+        # Always include shift/scale with nonlinear transformation
```
Comment (nirum) [inline twpca/model.py]: add support for `identity` warptype

**Review by nirum — CHANGES_REQUESTED** [review]

File: twpca/model.py [hunk 5]
```diff
@@ -89,117 +248,50 @@ def fit(self, X, optimizer=tf.train.AdamOptimizer, niter=1000, lr=1e-3, sess=Non
...
-        # TODO: compute low rank factors using initial warps, not raw data
-        # TODO: rewrite so this works on NaNs
-        trial_init, time_init, neuron_init = utils.compute_lowrank_factors(np_X, self.n_components, self.fit_trial_factors, self.nonneg, self.last_idx)
-
-        # create tensorflow variables for factor matrices
-        self._raw_params['time'] = tf.Variable(time_init, name="time_factors")
-        self._raw_params['neuron'] = tf.Variable(neuron_init, name="neuron_factors")
-        if self.fit_trial_factors:
-            self._raw_params['trial'] = tf.Variable(trial_init, name="trial_factors")
+        # TODO: better session management
```
Comment (poolio) [inline twpca/model.py]: remove

File: twpca/model.py [hunk 6]
```diff
@@ -55,26 +86,154 @@ def __init__(self, n_components, shared_length=None,
...
+
+        # assign the warps and return
+        ops = []
+        _var, var = (self.tau_shift, self.tau_scale, self.tau), (shift, scale, tau)
+        for _v, v in zip(_var, var):
+            ops += [tf.assign(_v, tf.constant(v, dtype=tf.float32))]
+        return self._sess.run(ops)
 
-    def fit(self, X, optimizer=tf.train.AdamOptimizer, niter=1000, lr=1e-3, sess=None, progressbar=True):
+    def fit(self, optimizer=tf.train.AdamOptimizer, niter=1000, lr=1e-3, sess=None, progressbar=True):
```
Comment (poolio) [inline twpca/model.py]: remove sess arg

File: twpca/model.py (hunk 2 above)
Comment (ahwillia) [reply twpca/model.py]: Done for now. But let's make a wrapper class to handle this!

File: twpca/model.py (hunk 3 above)
Comment (ahwillia) [reply twpca/model.py:105]: 👍

File: twpca/model.py (hunk 1 above)
Comment (ahwillia) [reply twpca/model.py:266]: Done.

File: twpca/model.py (hunk 4 above)
Comment (ahwillia) [reply twpca/model.py]: Done.

File: twpca/model.py [hunk 7]
```diff
@@ -1,27 +1,35 @@
...
+from sklearn.decomposition import NMF, TruncatedSVD
 from tqdm import trange
 
 import tensorflow as tf
 from . import warp, utils
 from .regularizers import l2, curvature
 
 
-class TWPCA(BaseEstimator, TransformerMixin):
+# TODO - niru make a super class that handles graph and session management
```
Comment (ahwillia) [inline twpca/model.py]: @nirum - add your wrapper class here?

File: twpca/model.py [hunk 8]
```diff
@@ -54,152 +95,250 @@ def __init__(self, n_components, shared_length=None,
...
+            # warp parameters
+            self.tau = tf.get_variable('tau', shape=(self.n_trials, self.shared_length), dtype=tf.float32)
+            self.tau_shift = tf.get_variable('tau_shift', shape=(self.n_trials,), dtype=tf.float32)
+            self.tau_scale = tf.get_variable('tau_scale', shape=(self.n_trials,), dtype=tf.float32)
+            # learning rate for optimizers
+            self._lr = tf.placeholder(tf.float32, shape=[])
+
+            # sets up tensorflow variables for warps
+            _pos_tau = tf.nn.softplus(self.tau) / tf.log(2.0)
+            _warp = self.tau_scale[:, None] * tf.cumsum(_pos_tau, 1) + self.tau_shift[:, None]
```
Comment (poolio) [inline twpca/model.py]: why prefix all these local variables with underscores?

File: twpca/model.py [hunk 9]
```diff
@@ -33,18 +41,51 @@ def __init__(self, n_components, shared_length=None,
                 to the neuron and time factors (default: False)
```
Comment (poolio) [inline twpca/model.py]: missing docstring for center_taus

File: twpca/model.py [hunk 10]
```diff
@@ -54,152 +95,250 @@ def __init__(self, n_components, shared_length=None,
...
+        # create all tensorflow variables
+        with self._graph.as_default():
+            self._train_vars = {} # tensorflow variables that are optimized
+            # factor matrix parameters
+            self._train_vars['time'] = tf.get_variable('time_factors', shape=(self.n_timepoints, self.n_components))
+            self._train_vars['neuron'] = tf.get_variable('neuron_factors', shape=(self.n_neurons, self.n_components))
+            if self.fit_trial_factors:
+                self._train_vars['trial'] = tf.get_variable('trial_factors', shape=(self.n_neurons, self.n_components))
+            # warp parameters
+            self.tau = tf.get_variable('tau', shape=(self.n_trials, self.shared_length), dtype=tf.float32)
```
Comment (poolio) [inline twpca/model.py]: not super important, but why do we have these stored in both self._train_vars and self.tau (same for other params)

File: twpca/model.py [hunk 11]
```diff
@@ -54,152 +95,250 @@ def __init__(self, n_components, shared_length=None,
...
+            # Force mean intercept to be zero and min slope to be one
+            if center_taus:
+                mean_intercept = tf.reduce_mean(_warp[:, 0])
+                min_slope = tf.reduce_min(_warp[:, -1] - _warp[:, 0]) / (self.n_timepoints - 1)
+                _warp = (_warp - mean_intercept) / min_slope
+
+            # Force warps to be identical at origin idx
+            if origin_idx is not None:
+                pin = _warp - _warp[:, origin_idx][:, None] + origin_idx
+                _warp = tf.clip_by_value(pin, 0, self.n_timepoints - 1)
```
Comment (poolio) [inline twpca/model.py]: add a TODO(ben): investigate clipping/boundary conditions

File: twpca/model.py [hunk 12]
```diff
@@ -54,152 +95,250 @@ def __init__(self, n_components, shared_length=None,
...
+
+            # store the warping and inverse warping function
+            self._params = {'warp': _warp}
+            _args = [_warp, self.n_timepoints, self.shared_length]
```
Comment (poolio) [inline twpca/model.py]: why create this list?

File: twpca/crossval.py [hunk 13]
```diff
@@ -11,12 +11,11 @@
...
-def cross_validate(model, data, nfits, drop_prob, **fit_kw):
+def cross_validate(data, nfits, drop_prob, model_kw, fit_kw):
     """Runs cross-validation on TWPCA model.
 
     Args
     ---
-    model : TWPCA model instance
     data (ndarray) : data tensor, trials x time x neurons
     nfits (int) : number of cross-validation runs
     drop_prob (float) : probability of setting element of data to nan
```
Comment (poolio) [inline twpca/crossval.py:21]: add model_kw/fit_kw to docstring

File: twpca/crossval.py [hunk 14]
```diff
@@ -53,29 +52,30 @@ def cross_validate(model, data, nfits, drop_prob, **fit_kw):
...
-             'train_error': model._sess.run(model.recon_cost),
+             'train_error': model.recon_cost,
              'test_error': test_error,
              'obj_history': model.obj_history
             }
         )
 
         # clean-up tensorflow
-        sess.close()
+        model._sess.close()
```
Comment (poolio) [inline twpca/crossval.py:71]: you can also just do sess.close() here as it's a local variable

File: twpca/crossval.py [hunk 15]
```diff
@@ -53,29 +52,30 @@ def cross_validate(model, data, nfits, drop_prob, **fit_kw):
...
 
     return results
 
 def hyperparam_search(data, n_components, warp_scales, time_scales,
                       nfits=1, drop_prob=0.1, warp_reg=None, time_reg=None,
                       fit_kw=dict(lr=(1e-1, 1e-2), niter=(250, 500), progressbar=False),
-                      **model_kw):
+                      model_kw=dict()):
```
Comment (poolio) [inline twpca/crossval.py]: don't have a mutable as a kwarg, set it to None and then assign it dict() if it is None

File: twpca/crossval.py [hunk 16]
```diff
@@ -132,8 +132,11 @@ def hyperparam_search(data, n_components, warp_scales, time_scales,
     # run cross-validation for all specified hyperparameters
     for nc, ws, ts in zip(tqdm(n_components), warp_scales, time_scales):
 
-        model = TWPCA(nc, warp_regularizer=warp_reg(ws), time_regularizer=time_reg(ts), **model_kw)
-        _result = cross_validate(model, data, nfits, drop_prob, **fit_kw)
+        model_args = dict(n_components=nc,
```
Comment (poolio) [inline twpca/crossval.py:135]: `model_kwargs` as this is a dict?

File: twpca/model.py [hunk 17]
```diff
@@ -54,152 +90,249 @@ def __init__(self, n_components, shared_length=None,
...
+        self._train_vars['time'] = tf.Variable(tf.zeros((self.n_timepoints, self.n_components)))
+        self._train_vars['neuron'] = tf.Variable(tf.zeros((self.n_neurons, self.n_components)))
+        if self.fit_trial_factors:
+            self._train_vars['trial'] = tf.Variable(tf.zeros((self.n_neurons, self.n_components)))
+        # warp parameters
+        self.tau = tf.Variable(tf.zeros((self.n_trials, self.shared_length)))
+        self.tau_shift = tf.Variable(tf.zeros(self.n_trials))
+        self.tau_scale = tf.Variable(tf.zeros(self.n_trials))
+        # learning rate for optimizers
+        self._lr = tf.placeholder(tf.float32, shape=[])
```
Comment (poolio) [inline twpca/model.py]: add doc here (or maybe at the top of the file?) that class variables prefixed with _ are TensorFlow Tensors

File: twpca/model.py [hunk 18]
```diff
@@ -54,152 +90,249 @@ def __init__(self, n_components, shared_length=None,
...
+        else:
+            # trial i, time j, factor k, neuron n
+            self._pred = tf.einsum('ijk,nk->ijn', self._warped_time_factors, self._params['neuron'])
+
+        # objective function (note that nan values are zeroed out by self._mask)
+        self._recon_cost = tf.reduce_sum(self._mask * (self._pred - self._data)**2) / self.num_datapoints
+        self._objective = self._recon_cost + self._regularization
+
+        # initialize values for tensorflow variables
+        self.assign_train_op(optimizer)
```
Comment (poolio) [inline twpca/model.py]: rename this to `create_train_op`

File: twpca/model.py [hunk 19]
```diff
@@ -54,152 +90,249 @@ def __init__(self, n_components, shared_length=None,
...
+        self.assign_warps(warps, normalize_warps)
```
Comment (poolio) [inline twpca/model.py]: i would prefer `initialize_warps` and `initialize_factors` to `assign`

File: twpca/model.py [hunk 20]
```diff
@@ -54,152 +90,249 @@ def __init__(self, n_components, shared_length=None,
...
+        self.assign_factors()
+
+    def assign_train_op(self, optimizer):
+        """Assign the training operation
+        """
+        self._opt = optimizer(self._lr)
+        var_list = [v for k, v in self._train_vars.items() if k != 'warp'] + list(self._train_vars['warp'])
```
Comment (poolio) [inline twpca/model.py]: isn't this just the same as `self._train_vars.values()`?

File: twpca/model.py [hunk 21]
```diff
@@ -54,152 +90,249 @@ def __init__(self, n_components, shared_length=None,
...
+            trial_fctr = np.empty((data.shape[0], n_components), dtype=np.float32)
+            for k, trial in enumerate(data):
+                t = last_idx[k] # last index before NaN
+                trial_fctr[k] = np.diag(np.linalg.pinv(time_fctr[:t]).dot(trial[:t]).dot(Bpinv.T))
+            assignment_ops += [tf.assign(self._train_vars['neuron'], trial_fctr)]
+
+        # done initializing factors
+        return self._sess.run(assignment_ops)
+
+    def assign_warps(self, warps, normalize_warps=True):
```
Comment (poolio) [inline twpca/model.py]: `warps` should be `warps=None`

File: twpca/utils.py [hunk 22]
```diff
@@ -59,56 +59,10 @@ def stable_rank(matrix):
...
-    # least squares (see Kolda & Bader, 2009) for CP tensor decomposition.
-    else:
-        Bpinv = np.linalg.pinv(neuron_fctr)
-        trial_fctr = np.empty((data.shape[0], n_components), dtype=np.float32)
-        for k, trial in enumerate(data):
-            t = last_idx[k] # last index before NaN
-            trial_fctr[k] = np.diag(np.linalg.pinv(time_fctr[:t]).dot(trial[:t]).dot(Bpinv.T))
-        return trial_fctr, time_fctr, neuron_fctr
-
+    return np.log(np.exp(y + 1e-5) - 1)
```
Comment (poolio) [inline twpca/utils.py]: could you make this smaller, and maybe put it in the log, like:
`np.log(np.exp(y) - 1 + 1e-9)`

File: twpca/crossval.py (hunk 16 above)
Comment (ahwillia) [reply twpca/crossval.py:135]: its actually a mix of keyword and positional args...

File: twpca/crossval.py (hunk 14 above)
Comment (ahwillia) [reply twpca/crossval.py:71]: oops, I didn't intend for sess to exist. Fixed now, good catch.

**Pull request merged**